### PR TITLE
Fixes in Assets Pallet

### PR DESCRIPTION
**Pull request by shawntabrizi**
* Track `approvals` in `approve_transfer`, `cancel_approval`, `force_cancel_approval` and `transfer_approved`.
* Use witness data as an upper bound rather than exact match for asset destroy.
* Don't allow approvals when an asset is frozen.
* Approvals are returned when asset is destroyed.
* Update `NonTransfer` proxy

Fixes: https://github.com/paritytech/srlabs_findings/issues/87
Fixes: https://github.com/paritytech/srlabs_findings/issues/88
Fixes: https://github.com/paritytech/srlabs_findings/issues/89

### CODE REVIEW

File: frame/assets/src/lib.rs [hunk 1]
```diff
@@ -468,6 +468,10 @@ pub mod pallet {
 		///
 		/// Emits `Destroyed` event when successful.
 		///
+		/// NOTE: It can be helpful to first freeze an asset before destroying it so that you
+		/// can provide accurate witness information and prevent users from manipulating state
+		/// in a way that can make it harder to destroy.
+		///
```
Comment (emostov) [inline frame/assets/src/lib.rs:474]: Should we also note that for assets with many holders and approvals in storage the owner may need to call `burn` and `force_cancel_approval` so the weight of `destroy` can fit into a block?

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -481,7 +485,7 @@ pub mod pallet {
 			origin: OriginFor<T>,
 			#[pallet::compact] id: T::AssetId,
 			witness: DestroyWitness,
```
Comment (emostov) [inline frame/assets/src/lib.rs:487]: @shawntabrizi whats the point of having a witness?

**Comment by emostov** [comment]
LGTM

re https://github.com/paritytech/srlabs_findings/issues/89 bullet point 3, it would be nice a bonus to have a note that the `destroy` extrinsic may not be able to fit into a block and the owner my have to burn/remove approvals one-by-one prior in order to get it to fit.

**Review by emostov — APPROVED** [review]

File: frame/assets/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply frame/assets/src/lib.rs:487]: Witness data is used to estimate the weight needed to execute that extrinsic without needing to read storage to actually know how many elements there are.

**Review by gavofyork — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**